### ISSUE TITLE
ttnn.atanh_bw returns NaN for every zero incoming gradient (torch returns 0)

### ISSUE THREAD

**Issue by Lanochu** [tenstorrent/tt-metal#56378]
## Summary

`atanh_bw` unconditionally replaces its result with NaN wherever the incoming gradient is zero,
then restores `0` only where the *input* is also exactly zero. For `grad == 0` and any `input != 0`
inside the domain, the op returns NaN where `torch.atanh`'s backward returns `0.0`.

| input | grad | torch (golden) | ttnn.atanh_bw | |
|---:|---:|---:|---:|---|
| 0.5 | 0.0 | 0.000000 | **nan** | wrong |
| 0.9 | 0.0 | 0.000000 | **nan** | wrong |
| -0.5 | 0.0 | 0.000000 | **nan** | wrong |
| 0.0 | 0.0 | 0.000000 | 0.000000 | ok (special-cased) |
| 0.5 | 1.0 | 1.333333 | 1.333333 | ok |
| 0.5 | -2.0 | -2.666667 | -2.666667 | ok |
| 1.0 | 0.0 | nan | nan | ok |
| 1.0 | 1.0 | inf | inf | ok |

Zero incoming gradients are not an edge case in training — they arise from masked losses, padding
tokens, `where`/`clamp` branches, dead units and sparse objectives. A NaN there propagates through
the rest of backprop and destroys the update, so this is a silent training-corruption bug rather
than an accuracy issue.

The golden is `torch.atanh` autograd (`ttnn/ttnn/operations/unary_backward.py#L136-L141`, via
`_golden_function_unary_backward`). A standalone reproduction (torch only, no device) is included
at the end of this issue.

## Root cause

`ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L962-L1001`:

```cpp
Tensor grad_a = ttnn::multiply(grad, unary_chain(input, ops_chain, ...));   // grad * 1/(1-x^2)
grad_a = where(ttnn::eqz(grad, ...), t_nan, grad_a, ...);                   // <-- every zero grad -> NaN
grad_a = where(
    ttnn::logical_and(ttnn::eqz(grad, ...), ttnn::eqz(input, ...)),
    0.f, grad_a, ...);                                                      // <-- restores only input == 0
```

The guard is evidently meant for the poles: at `|x| = 1` the derivative `1/(1-x^2)` is infinite, and
`0 * inf` is NaN, which is what torch produces. But the condition tests only `grad == 0` and ignores
where the input is, so it fires across the entire domain. The follow-up line then rescues a single
point, `input == 0`, which is presumably why the behaviour was believed correct.

This is the only occurrence of the pattern — I grepped `where(ttnn::eqz(grad` across
`unary_backward.cpp` and `atanh_bw` is the sole match, so no other backward op is affected.

## Why tests miss it

The existing test *does* feed zero gradients, and still cannot fail.

- `tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_atanh.py` builds both
  tensors with `data_gen_with_range(shapes, -100, 100)`, i.e.
  `torch.rand(shape).bfloat16() * 200 - 100`. Because the uniform sample is rounded to **bfloat16
  before** the affine map, `rand` lands on exactly `0.5` often, and `0.5 * 200 - 100` is exactly
  `0.0`:

  | shape | elements | exactly-zero grads |
  |---|---:|---:|
  | (1, 1, 32, 32) | 1,024 | 5 (0.488%) |
  | (1, 1, 320, 384) | 122,880 | 345 (0.281%) |
  | (1, 3, 320, 384) | 368,640 | 1,102 (0.299%) |

- The assertion is `compare_pcc` -> `comp_pcc` -> `get_atol_rtol_pcc`
  (`ttnn/tt_lib/_internal/comparison_funcs.py#L41-L55`), which contains:

  ```python
  # For now, mask all infs and nans so that we check the rest... TODO
  golden[isnan | isinf | isneginf] = 0
  calculated[isnan | isinf | isneginf] = 0
  ```

  The device's spurious NaN is replaced by `0` — which is exactly the value the golden already
  holds at those elements. **The mask maps the wrong answer onto the right answer**, so the
  comparison is not merely insensitive here, it is actively self-cancelling.

- There is no `tests/sweep_framework` entry for `atanh_bw` at all.

## Proposed fix

Restrict the NaN to the poles, where it is genuinely `0 * inf`:

```cpp
Tensor grad_a = ttnn::multiply(grad, unary_chain(input, ops_chain, ...));
// 0 * inf is NaN only at |x| = 1; for any other input a zero incoming gradient must give 0.
grad_a = where(
    ttnn::logical_and(
        ttnn::eqz(grad, ...),
        ttnn::eq(ttnn::abs(input, ...), 1.0f, std::nullopt, ...), ...),
    t_nan, grad_a, ...);
```

and delete the `eqz(grad) && eqz(input)` line, which becomes redundant. The remaining
`|input| == 1 && nez(grad) -> ±inf` handling is already correct and should stay.

Strictly, deleting both lines outright also gives the right answers under IEEE semantics, since
`grad * recip(0) = 0 * inf = NaN` at the poles and `0 * finite = 0` elsewhere; the explicit form
above is proposed only because it does not depend on what `ttnn::reciprocal(0)` returns.

For tests I would add a zero-gradient case asserted with an exact comparison rather than PCC, and
separately note that the NaN/inf masking in `get_atol_rtol_pcc` (already marked `TODO`) will hide
this entire class of bug in any backward op.

## Validation status

The device behaviour above is a transcription of the composite op's `where` chain, not a silicon
run; the torch column is real `torch.atanh` autograd output. The defect is visible directly in the
source and does not depend on floating-point modelling. I have not run this on a Wormhole or
Blackhole part and would want on-device confirmation as the first step of the fix.

<details>
<summary><b>Standalone reproduction</b> (torch only, no Tenstorrent hardware required)</summary>

```python
"""
Reproduction for ttnn.atanh_bw returning NaN on zero incoming gradients.

The 'ttnn' column is a transcription of the `where` chain in
ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp::atanh_bw.
The 'torch' column is real autograd output -- the registered golden.

Requires: torch.   Run:  python atanh_bw_repro.py
"""
import torch

def ttnn_atanh_bw(x, g):
    """Transcription of unary_backward.cpp::atanh_bw."""
    ga = g * (1.0 / -(x * x - 1.0)) if abs(x) != 1.0 else float("nan")
    if g == 0.0:                                ga = float("nan")   # where(eqz(grad), NaN, ...)
    if g == 0.0 and x == 0.0:                   ga = 0.0            # where(eqz(grad) & eqz(input), 0, ...)
    if abs(x) == 1.0 and g != 0.0:              ga = float("inf")   # where((|x|==1) & nez(grad), inf, ...)
    if ga == float("inf") and g < 0:            ga = float("-inf")
    return ga

print("1) Zero incoming gradient\n")
print(f"   {'input':>7} {'grad':>6} {'torch (golden)':>16} {'ttnn':>10}   verdict")
for x0, g0 in [(0.5,0.0),(0.9,0.0),(-0.5,0.0),(0.0,0.0),(0.5,1.0),(0.5,-2.0),(1.0,0.0),(1.0,1.0)]:
    t = torch.tensor([x0], dtype=torch.float64, requires_grad=True)
    torch.atanh(t).backward(torch.tensor([g0], dtype=torch.float64))
    ref, got = t.grad.item(), ttnn_atanh_bw(x0, g0)
    same = (ref != ref and got != got) or ref == got
    print(f"   {x0:7.2f} {g0:6.1f} {ref:16.6f} {got:10.6f}   {'ok' if same else 'MISMATCH'}")

print("\n2) The existing test's own data contains exact zero gradients\n")
print("   test_backward_atanh.py uses data_gen_with_range(shape, -100, 100):")
print("   torch.rand(shape).bfloat16() * 200 - 100  -- rand rounds to exactly 0.5 -> grad exactly 0\n")
torch.manual_seed(0)
for shape in [(1,1,32,32),(1,1,320,384),(1,3,320,384)]:
    t = torch.rand(shape).bfloat16() * 200 - 100
    n = int((t == 0).sum())
    print(f"   shape {str(shape):<16} elements={t.numel():>8,}  exactly-zero grads={n:>6,} ({n/t.numel():.3%})")

print("\n3) Why the test still passes\n")
print("   comp_pcc -> get_atol_rtol_pcc (ttnn/tt_lib/_internal/comparison_funcs.py:41) does:")
print("       golden[isnan|isinf] = 0 ;  calculated[isnan|isinf] = 0")
print("   The device's NaN becomes 0 -- exactly the value the golden already holds there.")
print("   The mask maps the wrong answer onto the right answer, so the check cannot fail.")

```

</details>

**Comment by Lanochu** [comment]
I have a fix for this ready and would like to be assigned before opening a PR, per the bounty
terms ("you must be assigned on GitHub to the issue for which you are submitting a pull request").

**The change** — restrict the NaN to the poles, where `0 * inf` genuinely is NaN, instead of
testing `grad` alone:

```cpp
// A zero incoming gradient must propagate as zero. NaN is correct only at the poles
// |input| == 1, where the local derivative 1/(1-x^2) is infinite and 0 * inf is NaN.
// Testing grad alone NaN'd every zero gradient and rescued only input == 0.
grad_a = where(
    ttnn::logical_and(
        ttnn::eqz(grad, output_mem_config),
        ttnn::logical_or(
            ttnn::eq(input, 1.0f, std::nullopt, output_mem_config),
            ttnn::eq(input, -1.0f, std::nullopt, output_mem_config),
            std::nullopt,
            output_mem_config),
        std::nullopt,
        output_mem_config),
    t_nan,
    grad_a,
    output_mem_config);
```

This replaces both the `eqz(grad) -> NaN` line and the now-redundant `eqz(grad) && eqz(input) -> 0`
line. The subsequent `|input| == 1 && nez(grad) -> ±inf` handling is already correct and is
untouched. Net diff is 13 added / 3 removed lines in `unary_backward.cpp`.

**Regression test** added to `test_backward_atanh.py` with an all-zero gradient across
`{0.5, 0.9, -0.5, 0.0, 1.0, -1.0}`, compared against the registered golden with an exact
comparison rather than PCC — as described in the issue, the NaN/inf masking in
`get_atol_rtol_pcc` maps the device's spurious NaN onto `0`, which is exactly the value the golden
holds there, so a PCC assertion cannot fail on this.

Verified against `torch.atanh` autograd on a 14-point grid spanning the interior, both poles and
zero/nonzero gradients: all 14 match after the change (before it, every `grad == 0, input != 0`
case is NaN against a golden of 0). As noted in the issue, I have not built tt-metal locally — every
construct in the patch already appears inside this same function, but your CI would be the first
real compile.

Happy to open the PR as soon as it is assigned.

**Comment by Lanochu** [comment]
Closing as a duplicate of #55290, which @ping-dev-ui filed on 2026-09-02 and which identifies the
same two lines and the same root cause — the `eqz(grad)` guard was meant to encode `0 * inf = NaN`
at the poles but carries no pole condition. Apologies for the noise; I should have searched before
filing.

**Comment by Lanochu** [comment]
Closing as a duplicate of #55290 (ping-dev-ui, 2026-09-02), which reports exactly this — the `where(eqz(grad), t_nan, grad_a)` guard having no pole condition, with only `x == 0` rescued by the second guard.

My prior-art search used `atanh_bw in:title`, and GitHub's search did not return #55290 for that query despite the term being in its title — underscored identifiers tokenize unreliably. That is my error, not a defect in #55290. Apologies for the duplicate.
